system/lib/wasmfs/file.h

```diff
@@ -8,37 +8,133 @@
 
 #pragma once
 
+#include <assert.h>
 #include <emscripten/html5.h>
 #include <mutex>
+#include <unordered_map>
 #include <vector>
 #include <wasi/api.h>
 
 namespace wasmfs {
 
-class File {
-  // TODO: Add other File properties later.
+class File : public std::enable_shared_from_this<File> {
+
+public:
+  enum FileKind { DataFileKind = 0, DirectoryKind, SymlinkKind };
+
+  // curr->is<Directory>()
+  template <class T> bool is() const {
+    static_assert(std::is_base_of<File, T>::value, "File is not a base of destination type T");
+    return int(kind) == int(T::SpecificFileKind);
+  }
+
+  template <class T> T* dynCast() {
+    static_assert(std::is_base_of<File, T>::value, "File is not a base of destination type T");
+    return int(kind) == int(T::SpecificFileKind) ? (T*)this : nullptr;
+  }
+
+  template <class T> const T* dynCast() const {
+    static_assert(std::is_base_of<File, T>::value, "File is not a base of destination type T");
+    return int(kind) == int(T::SpecificFileKind) ? (const T*)this : nullptr;
+  }
+
+  template <class T> T* cast() {
+    static_assert(std::is_base_of<File, T>::value, "File is not a base of destination type T");
+    assert(int(kind) == int(T::SpecificFileKind));
+    return (T*)this;
+  }
+
+  template <class T> const T* cast() const {
+    static_assert(std::is_base_of<File, T>::value, "File is not a base of destination type T");
+    assert(int(kind) == int(T::SpecificFileKind));
+    return (const T*)this;
+  }
+
+  class Handle {
+
+    std::unique_lock<std::mutex> lock;
 
+  protected:
+    std::shared_ptr<File> file;
+
+  public:
+    Handle(std::shared_ptr<File> file) : file(file), lock(file->mutex) {}
+  };
+
+  Handle get() { return Handle(shared_from_this()); }
+
+protected:
+  File(FileKind kind) : kind(kind) {}
   // A mutex is needed for multiple accesses to the same file.
   std::mutex mutex;
 
+private:
+  // TODO: Add other File properties later.
+
+  FileKind kind;
+};
+
+class DataFile : public File {
+
   virtual __wasi_errno_t read(const uint8_t* buf, __wasi_size_t len) = 0;
   virtual __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) = 0;
```

Comment:
These methods should probably move to `DataFile::Handle`, since they don't make sense on other kinds of files.